bob/playbook/rebuild.go

```diff
@@ -5,77 +5,111 @@ import (
 	"fmt"
 
 	"github.com/benchkram/bob/bobtask"
+	"github.com/benchkram/bob/bobtask/target"
 	"github.com/benchkram/bob/pkg/boberror"
 	"github.com/benchkram/bob/pkg/boblog"
 	"github.com/benchkram/bob/pkg/usererror"
 	"github.com/benchkram/errz"
 )
 
+// RebuildInfo contains information about a task rebuild: if it's required and the cause for it
+type RebuildInfo struct {
+	// IsRequired tells if the task requires rebuild again
+	IsRequired bool
+	// Cause tells why the rebuild is required
+	Cause RebuildCause
+	// VerifyResult is the result of target filesystem verification
+	VerifyResult target.VerifyResult
+}
+
 // TaskNeedsRebuild check if a tasks need a rebuild by looking at its hash value
 // and its child tasks.
-func (p *Playbook) TaskNeedsRebuild(taskname string) (rebuildRequired bool, cause RebuildCause, err error) {
-	ts, ok := p.Tasks[taskname]
+func (p *Playbook) TaskNeedsRebuild(taskName string) (rebuildInfo RebuildInfo, err error) {
+	ts, ok := p.Tasks[taskName]
 	if !ok {
-		return false, "", usererror.Wrap(boberror.ErrTaskDoesNotExistF(taskname))
+		return RebuildInfo{}, usererror.Wrap(boberror.ErrTaskDoesNotExistF(taskName))
 	}
 	task := ts.Task
 	coloredName := task.ColoredName()
 
 	// Rebuild strategy set to `always`
 	if task.Rebuild() == bobtask.RebuildAlways {
 		boblog.Log.V(3).Info(fmt.Sprintf("%-*s\tNEEDS REBUILD\t(rebuild set to always)", p.namePad, coloredName))
-		return true, TaskForcedRebuild, nil
+
+		// for forced rebuild all task targets are marked as invalid
+		invalidFiles := make(map[string][]target.Reason)
+		if task.TargetExists() {
+			t, err := ts.Target()
+			errz.Fatal(err)
+			invalidFiles = t.AsInvalidFiles(target.ReasonForcedByNoCache)
+		}
+
+		return RebuildInfo{IsRequired: true, Cause: TaskForcedRebuild, VerifyResult: target.VerifyResult{
+			TargetIsValid: len(invalidFiles) > 0,
+			InvalidFiles:  invalidFiles,
+		}}, nil
 	}
 
 	// Did a child task change?
-	if p.didChildTaskChange(task.Name(), p.namePad, coloredName) {
+	if p.didChildTaskChange(task.Name()) {
 		boblog.Log.V(3).Info(fmt.Sprintf("%-*s\tNEEDS REBUILD\t(dependecy changed)", p.namePad, coloredName))
-		return true, DependencyChanged, nil
+		return RebuildInfo{IsRequired: true, Cause: DependencyChanged}, nil
```

Comment:
A changing dependency can invalidate a target. Should probably also passed using `AsInvalidFIles`

Reply:
Indeed is possible. How to reproduce:

Have 2 tasks like:

```yaml
  withChangingDependency:
    cmd: touch unchanged
    target: unchanged
    dependsOn:
      - shouldChange
  shouldChange:
    cmd: echo hello
```

On `bob build withChangingDependency` it creates an empty `unchanged` file and subsequent builds will be marked as cached.

Change the `shouldChange` task to invalidate target of `withChangingDependency`.

```yaml
  withChangingDependency:
    cmd: touch unchanged
    target: unchanged
    dependsOn:
      - shouldChange
  shouldChange:
    cmd: echo hello > unchanged
```

Running `bob build withChangingDependency` will rebuild both tasks but the content of the `unchanged` file will be hello instead of empty(it should be brought back from the cache).

Fix applied: If child task has changed check the target of the task and add corresponding invalidFiles to verifyResult.